### PR TITLE
ZK-Lambda-Class Remade

### PR DESCRIPTION
**Pull request by Ghilker**
### This is the same as #51159 with the requested changes done, there may be slip-ups so let me know. Remade because of problems with the other branch

This PR adds the crystal invasion event, a new event involving the Supermatter and the monsters from within. 
When the event starts there will be a message from centcomm announcing it, then the supermatter explode leaving a destabilized crystal that emits radiations and harmful gases; after a bit portals will spawn around the station, that will produce a number of monsters each. there are 4 types of waves and 4 types of portals
The waves are: Small, Medium, Big, Huge (each have different kind of portals that can spawn and different amounts too)
The portals are the same types of the waves, they differ from each other for the number of monsters that can spawn and the kind of monsters that can spawn (bigger portals spawn stronger monsters)
To end the event the players should stabilize the crystal by destroying the portals (for now are indestructible and they are disabled by using an anomaly neutralizer, might change that) and collecting _anomaly cores_ **otherworld  crystals**; then those _cores_ **crystals** are to be put in the crystal stabilizer, an item unlockable in the tech tree. After this just inject the destabilized crystal with it and the remaining portals will close on their own (the spawned monsters will still remain tho so you have to slay them)

All the numbers are mostly eyeballed and could change if requested/with feedbacks
Code Docs will come soon!
## Why It's Good For The Game
New event and crew content is always good

## Changelog
:cl:
add: Added Crystal Invasion Event (with all the mechanics, mobs and item related)
/:cl:

### CODE REVIEW

**Comment by optimumtact** [comment]
Please switch away from using anomaly cores, they're reserved for control of high level tech items and are bound by the document on the hackmd.

**Comment by LemonInTheDark** [comment]
It lives :)

**Comment by Ghilker** [comment]
changed the drop for closing the portals from an anomaly core to a crystal

**Comment by trerri** [comment]
Would be cool if the mobs had some mechanical differences. For example, the tiny mob deals burn when you hit it but dies in one hit to thrown items, the small mob explodes into limbs that fly off and really sting when they hit you, the big mob dusts you when touched and needs guns to be killed, the huge mob is that with a hitscan laser

File: code/modules/events/crystal_event.dm [hunk 1]
```diff
@@ -0,0 +1,509 @@
...
+	announceWhen = 1
+	endWhen = 25 MINUTES
+	///Is the name of the wave, used to check wich wave will be generated
+	var/wave_name
+	///Max number of portals that can spawn per type of wave
+	var/portal_numbers
+	///Check if this is the first wave or not
+	var/spawned = FALSE
+	///Store the destabilized crystal
+	var/obj/machinery/destabilized_crystal/dest_crystal
```
[the remark this answers is not in the document]
Comment (Rohesie) [reply code/modules/events/crystal_event.dm:48]: Once you set up the `dest_crystal` (with stuff like `dest_crystal = thing`) register to its destruction and handle it appropriately.
For example: `RegisterSignal(dest_crystal, COMSIG_PARENT_QDELETING, .proc/on_dest_crystal_qdel)`
Clear the variable on `proc/on_dest_crystal_qdel`, unregister to it and do whatever makes sense to.

**Comment by Rohesie** [comment]
Alright, handle that last change and we'll ask to get this TMed

**Comment by Ghilker** [comment]
should be handled now

**Comment by Rohesie** [comment]
Poke the keyholders to get it TMed.

File: code/datums/materials/basemats.dm [hunk 2]
```diff
@@ -347,6 +347,16 @@ Unless you know what you're doing, only use the first three numbers. They're in
 	M.apply_damage(15, BRUTE, BODY_ZONE_HEAD, wound_bonus = 7)
 	return TRUE
 
+/datum/material/otherworld_crystal
+	name = "Otherworld Crystal"
+	desc = "An unkown crystal from an unkown dimension"
+	color = "#6d7e8e"
+	strength_modifier = 1.5
```
Comment (Qustinnus) [inline code/datums/materials/basemats.dm:354]: way too high. try to refrain from making materials this strong

Comment (Qustinnus) [reply code/datums/materials/basemats.dm:354]: generally if it's close to the strength of adamantine, its way too high.

File: code/modules/events/crystal_event.dm [hunk 3]
```diff
@@ -0,0 +1,517 @@
...
+
+/*
+This section is for the crystal portals variations
+*/
+/obj/structure/crystal_portal
+	name = "crystal portal"
+	desc = "this shouldn't be here"
+	icon = 'icons/effects/effects.dmi'
+	icon_state = "anom"
+	color = "#B2FFFE"
```
Comment (Rohesie) [inline code/modules/events/crystal_event.dm:306]: Missed this one. Use a color define. Check `colors.dm`
Add one if necessary.

File: code/modules/events/crystal_event.dm [hunk 4]
```diff
@@ -0,0 +1,517 @@
...
+		to_chat(user, "<span class='notice'>You start closing \the [src]...</span>")
+		if(W.use_tool(src, user, 5.5 SECONDS, volume = 100))
+			to_chat(user, "<span class='notice'>You successfully close \the [src]!</span>")
+			closed = TRUE
+			qdel(src)
+
+/obj/structure/crystal_portal/small
+	name = "Small Portal"
+	desc = "A small portal to an unkown dimension!"
+	color = "#B2FFFE"
```
Comment (Rohesie) [inline code/modules/events/crystal_event.dm:363]: Define

File: code/modules/events/crystal_event.dm [hunk 5]
```diff
@@ -0,0 +1,517 @@
...
+	spawn_time = 200
+	mob_types = list(
+		/mob/living/simple_animal/hostile/crystal_monster/minion,
+		/mob/living/simple_animal/hostile/crystal_monster/thug
+		)
+
+/obj/structure/crystal_portal/medium
+	name = "Medium Portal"
+	desc = "A medium portal to an unkown dimension!"
+	color = "#93F3B2"
```
Comment (Rohesie) [inline code/modules/events/crystal_event.dm:374]: Define

File: code/datums/materials/basemats.dm [hunk 6]
```diff
@@ -347,6 +347,16 @@ Unless you know what you're doing, only use the first three numbers. They're in
...
+	categories = list(MAT_CATEGORY_RIGID = TRUE, MAT_CATEGORY_BASE_RECIPES = FALSE)
```
Comment (Qustinnus) [inline code/datums/materials/basemats.dm:355]: don't set it to false just remove it from the list
```suggestion
	categories = list(MAT_CATEGORY_RIGID = TRUE)
```

**Comment by Rohesie** [comment]
Let me know when all the change requests are done so I can slap the TM tag again.

**Comment by Ghilker** [comment]
they done lmao

**Comment by Doosteroni** [comment]
I like how skellys are denoted in rank with the ratio of cheddar to bone, with the cheesiest being the best

**Comment by Ghilker** [comment]
> I like how skellys are denoted in rank with the ratio of cheddar to bone, with the cheesiest being the best

my gf saw this and said that she adores you

**Review by Twaticus — APPROVED** [review]
cheese skele

**Pull request merged**